spot-contracts/contracts/RolloverVault.sol

```diff
@@ -394,6 +400,118 @@ contract RolloverVault is
         return redemptions;
     }
 
+    /// @notice Allows users to swap their underlying tokens for perps held by the vault
+    /// @param underlyingAmtIn The amount of underlying tokens swapped in.
+    /// @param execRecover If set to true, recovers vault assets to maximize liquidity for swapping.
+    /// @return The amount of perp tokens swapped out.
+    function swapUnderlyingForPerps(uint256 underlyingAmtIn, bool execRecover)
+        public
+        nonReentrant
+        whenNotPaused
+        returns (uint256)
+    {
+        // compute equal value perps to swap out
+        uint256 perpPrice = perp.getAvgPrice();
+        uint256 perpAmtOut = underlyingAmtIn.mulDiv(PERP_UNIT_PRICE, perpPrice);
+
+        // Compute fees
+        (uint256 swapFeePerpSharePerc, uint256 swapFeeVaultSharePerc) = feePolicy.computeUnderlyingToPerpSwapFeePercs(
+            underlyingAmtIn
+        );
+
+        // We compute perp's fee share
+        uint256 perpAmtToBurn = perpAmtOut.mulDiv(swapFeePerpSharePerc, FEE_ONE_PERC, MathUpgradeable.Rounding.Up);
+
+        // We deduct fees by transferring out fewer perp tokens
```

Comment:
There are two options: we could transfer out fewer perps, or we could withhold AMPL before minting perps to begin with.

I think the second one would require less gas and seems easier, wdyt? Also, potentially leaves the vault with fewer perps that it'll probably end up reclaiming anyway. The vault generally prefers to hold AMPL over perps.

Reply:
Gas-wise I don't think it should make any difference as it will involve the same number of operations.

It becomes tricky as there are 2 parts to the fee. The vault share which could be held in either ampl or perps and the perp share which can only be paid by burning perps. It seems cleaner to have both parts be paid in the same token, if possible.

Reply:
Let's try the second one, since it doesn't seem much more complicated. This should save some extra recover calls.

Reply:
But wait, I think we're already doing the 2nd one?
Say user transfers 100 AMPL and the fees for perp/vault are 2 and 3 % respectively. 

perpAmt to burn is 2 perps, perpAmt to the user is 95. We first transfer 100 AMPL from the user, then mint 97 perps out of which 95 are transferred to the user and 2 are burnt.

Rather, we don't mint 100 perps and keep 3 as the vault fee.

Reply:
Oh you're right! Yeah this is great 👍